pkg/engine/task/task_pipe.go

```diff
@@ -0,0 +1,427 @@
+			return fmt.Errorf("pipe file name %s should only contain alphanumeric characters, '.', '_' and '-'", fileName)
+		}
+	}
+	return nil
+}
+
+func pipePod(ctr *corev1.Container, name string) (string, error) {
+	volumeName := ctr.VolumeMounts[0].Name
+	mountPath := ctr.VolumeMounts[0].MountPath
+
+	pod := corev1.Pod{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Pod",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Spec: corev1.PodSpec{
+			Volumes: []corev1.Volume{
+				{
+					Name:         volumeName,
+					VolumeSource: corev1.VolumeSource{EmptyDir: nil},
+				},
+			},
+			InitContainers: []corev1.Container{*ctr},
+			Containers: []corev1.Container{
+				{
+					Name:    pipePodContainerName,
+					Image:   "busybox",
+					Command: []string{"/bin/sh", "-c"},
+					Args:    []string{"sleep infinity"},
+					VolumeMounts: []corev1.VolumeMount{
+						{
+							Name:      volumeName,
+							MountPath: mountPath,
+						},
+					},
+				},
+			},
+			RestartPolicy: corev1.RestartPolicyOnFailure,
+		},
+	}
+
+	b, err := yaml.Marshal(pod)
+	if err != nil {
+		return "", fmt.Errorf("failed to create pipe task pod: %v", err)
+	}
+
+	return string(b), nil
+}
+
+func copyFiles(fs afero.Fs, ff []PipeFile, pod *corev1.Pod, ctx Context) error {
+	restCfg, err := config.GetConfig()
+	if err != nil {
+		return fatalExecutionError(fmt.Errorf("failed to fetch cluster REST config: %v", err), pipeTaskError, ctx.Meta)
+	}
+
+	var g errgroup.Group
+
+	for _, f := range ff {
+		f := f
+		log.Printf("PipeTask: %s/%s copying pipe file %s", ctx.Meta.InstanceNamespace, ctx.Meta.InstanceName, f.File)
+		g.Go(func() error {
+			return copyFile(fs, f, pod, restCfg)
+		})
+	}
+
+	err = g.Wait()
+	return err
+}
+
+func copyFile(fs afero.Fs, pf PipeFile, pod *corev1.Pod, restCfg *rest.Config) error {
+	reader, stdout := io.Pipe()
+
+	defer reader.Close()
+	defer stdout.Close()
+
+	pe := PodExec{
+		RestCfg:       restCfg,
+		PodName:       pod.Name,
+		PodNamespace:  pod.Namespace,
+		ContainerName: pipePodContainerName,
+		Args:          []string{"tar", "cf", "-", pf.File},
+		In:            nil,
+		Out:           stdout,
+		Err:           nil,
+	}
+
+	if err := pe.Run(); err != nil {
+		return fmt.Errorf("failed to copy pipe file. err: %v", err)
+	}
+
+	if err := untarFile(fs, "", reader, pf.File); err != nil {
+		return fmt.Errorf("failed to untar pipe file: %v", err)
+	}
+
+	return nil
+}
+
+// pipeFiles iterates through passed pipe files and their copied data, reads them, constructs k8s artifacts
+// and marshals them.
+func pipeFiles(fs afero.Fs, files []PipeFile, ctx Context) (map[string]string, error) {
+	artifacts := map[string]string{}
+
+	for _, pf := range files {
+		data, err := afero.ReadFile(fs, pf.File)
+		if err != nil {
+			return nil, fmt.Errorf("error opening pipe file %s", pf.File)
+		}
+
+		// API server has a limit of 1Mb for Secret/ConfigMap
+		if len(data) > 1024*1024 {
```

Comment:
should this be part of some kind of validation?

Reply:
see Andreases comment below